storyblok/field-plugin
# feat(lib): add modalSize object to the toggleModal message

## What?

- Retrieve `isAIEnabled` as plugin state
- Add the `requestUserContext` action and postMessage functionality to grab the user info `isSpaceAdmin` and `permissions`
- Add the `modalSize` as an optional parameter in the `toggleModal` postMessage to allow setting the `max-width` and/or `max-height` of the new modal with portal. (Only applicable when `enablePortalModal` is equal to `true`.

## Why?

JIRA: SHAPE-7911 & SHAPE-7912

The user info, `isSpaceAdmin` is needed by us and the `permissions` property was requested on this [issue](https://github.com/storyblok/field-plugin/issues/424) and also needed in some of our integrations as, for example, the Akeneo PIM.

The need to define custom modal sizes and the `isAIEnabled` state property is an internal demand of an integration we are currently working on.

## How to test? (optional)

To test it out, you will need to:

1. Run the Storyfront locally using the `feat/SHAPE-7911-get-the-new-provided-values-and-define-the-new-action` branch (still being reviewed https://github.com/storyblok/storyfront/pull/6904).

2. Open the `packages/demo/src/components/FieldPluginDemo.tsx` file and specify the `targetOrigin` to `http://localhost:3300` (Storyfront running address):
![image](https://github.com/user-attachments/assets/d211cba8-357a-4bde-a627-661c8bb98adc)

3. [Create a new fieldtype](http://localhost:3300/#/me/plugins) and deploy the field plugin demo:
```sh
yarn
yarn build
yarn workspace demo build
```
Now, copy the generated bundle file (located at `packages/demo/dist/index.js`) and paste it into the newly created field.

4. Add the new field type to any of your stories.
If everything was configured properly, you should see, in your Visual Editor, something similar to:

<img width="244" alt="Screenshot 2025-01-02 at 18 39 25" src="https://github.com/user-attachments/assets/31f3d071-97c0-4f31-bced-151c6ce3312c" />

### Test new Actions:

#### New modal sizing

Check the modal sizing by clicking on the `Toggle Modal` button and check if the modal max-width [was set to '50%'](https://github.com/storyblok/field-plugin/pull/447/files#diff-191feb44a592c69f3cab460e02149f93fe264a878dba3521fa3d853d1e89547dR13) accordingly. Try changing this value and also adding a `height`.

<img width="910" alt="Screenshot 2025-01-03 at 12 07 07" src="https://github.com/user-attachments/assets/d22e4494-4d02-4c7c-86ae-8a39a3580734" />

#### Request user context

You can change the logged-in user's permissions (under the space's Settings page) and see them reflected in the displayed text when requesting the user context by clicking the "Request user context" button.
<img width="925" alt="Screenshot 2025-01-02 at 18 42 14" src="https://github.com/user-attachments/assets/25898175-2066-4aee-823b-6705102117a9" />

Then, access the story with a non-admin user to see if the `isSpaceAdmin` property is set to "No" when the user context is requested.

### Test new State

Disable/Enable the AI feature under the space's settings page and check if the `isAIEnabled` property is correct.
![image](https://github.com/user-attachments/assets/b6edfa67-e2c0-419a-865f-3ea6102685c9)
